### DISCUSSION TITLE
How to write a controller for ROS2

### DISCUSSION THREAD

**Discussion by Jokua** [Monash-Connected-Autonomous-Vehicle/autoware#36] [Q&A]
Hi, I'm from the software team on the ESDA section. I wanted to ask if anyone can explain to me how do I write a controller for ros2 because the tutorial in this [link](https://control.ros.org/humble/doc/ros2_controllers/doc/writing_new_controller.html) isn't making much sense to me after the part for making the package directory. I would mention that I am clueless when it comes to C++ so bear with me on this one, plus I may have been looking up the wrong stuff so some clarification and pointers would be appreciated. Thanks!

**Comment by Jokua** [comment]
Hi @Monash-Connected-Autonomous-Vehicle/avp-team, forgot to tag you guys for this discussion but yeah anyone who understands C++ can you guys tell me what this error is and how to fix it?

This is from a ros2-controller package that i found online that has an existing ackermann steering controller implementation, but i got these errors which got me stumped. These are from the .hpp file
![image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/05df1947-ef3b-4fd6-b154-0a50f4d761d6)

![image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/68219577-d3fe-4d35-b515-83f278d2c029)

The second error i found the file in the repo, but I'm not entirely sure why it is coming up as an error.
![image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/1663db36-ce28-42b1-9399-dba694eec8df)

This issue also extends to the .cpp file for the controller as well, so it's consistent that i'm figuring it's due to something not properly setup. I haven't worked with C++ much before, so please bear with me being a noob LOL. Thanks!

P.S: Here's the repo link to the [ros2-controller page](https://github.com/ros-controls/ros2_controllers) and the [ros2-control page](https://github.com/ros-controls/ros2_control?tab=readme-ov-file) which may be related

**Reply by dylan-gonzalez** [reply]
Sorry, didn't see this until you tagged the team

`ackermann_steering_controller_parameters.hpp` is generated from [generate_parameters_library](https://github.com/PickNikRobotics/generate_parameter_library) at build time, and you can find it in the `lib` directory inside the workspace `install/` directory. 

I would probably ignore these errors, unless you have properly set up ros2 with vscode, which idk how to, but not important.

Have you actually tried running the ackermann_steering_controller? You also shouldn't have to clone the ros2_controllers repo, you can use the [binary packages](https://control.ros.org/humble/doc/getting_started/getting_started.html#binary-packages). You can test it out existing controllers by running the examples in the [ros2_control_demos repo](https://github.com/ros-controls/ros2_control_demos). 

(Also side note, but it would be more useful if you asked these sort of questions in the [issue](https://github.com/orgs/Monash-Connected-Autonomous-Vehicle/projects/11/views/1?pane=issue&itemId=49274385) you're working on, and should be able to ping the avp-team if needed there.

**Reply by Jokua** [reply]
Ah ok, i'll give it a go. Thanks!
